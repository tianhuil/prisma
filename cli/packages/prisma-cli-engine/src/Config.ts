import { RunOptions } from './types/common'
import * as path from 'path'
import * as os from 'os'
import * as fs from 'fs-extra'
import * as cuid from 'scuid'
import * as findUp from 'find-up'
import { Output } from './Output/index'
import { getGraphQLConfig } from 'graphql-config'
import { values } from 'lodash'
import { getRoot } from './util'

const isDevConsole =
  (process.env.CONSOLE_ENDPOINT || '').toLowerCase() === 'dev'

export class Config {
  /**
   * Local settings
   */
  mockInquirer?: any
  out: Output
  debug: boolean = Boolean(
    process.env.DEBUG && process.env.DEBUG!.includes('*'),
  )
  windows: boolean = false
  bin: string = 'prisma'
  mock: boolean = true
  argv: string[] = process.argv.slice(1)
  commandsDir: string = path.join(__dirname, '../dist/commands')
  defaultCommand: string = 'help'
  userPlugins: boolean = false
  version: string = '1.1'
  name: string = 'prisma'
  pjson: any = {
    name: 'cli-engine',
    version: '0.0.0',
    dependencies: {},
    'cli-engine': {
      defaultCommand: 'help',
    },
  }

  /**
   * Paths
   */
  cwd: string
  home: string
  root = path.join(__dirname, '..')

  definitionDir: string
  definitionPath: string | null
  globalPrismaPath: string
  globalConfigPath: string
  globalClusterCachePath: string
  warnings: string[] = []

  /**
   * Urls
   */
  cloudApiEndpoint =
    process.env.CLOUD_API_ENDPOINT || 'https://api2.cloud.prisma.sh'
  consoleEndpoint = isDevConsole
    ? 'http://localhost:3000'
    : 'https://app.prisma.io'

  /* tslint:disable-next-line */
  __cache = {}

  constructor(options?: RunOptions) {
    this.cwd = (options && options.cwd) || this.getCwd()
    this.home = (options && options.home) || this.getHome()
    this.setDefinitionPaths()
    this.setPaths()
    this.readPackageJson(options!)
    if (options && options.mockInquirer) {
      this.mockInquirer = options.mockInquirer
    }
  }
  setOutput(out: Output) {
    this.out = out
    this.warnings.forEach(warning => out.warn(warning))
    this.warnings = []
  }
  get arch(): string {
    return os.arch() === 'ia32' ? 'x86' : os.arch()
  }
  get platform(): string {
    return os.platform() === 'win32' ? 'windows' : os.platform()
  }
  get userAgent(): string {
    return `${this.name}/${this.version} (${this.platform}-${this.arch}) node-${
      process.version
    }`
  }
  get dirname() {
    return this.pjson['cli-engine'].dirname || this.bin
  }
  get cacheDir() {
    const x = dir(
      this,
      'cache',
      this.platform === 'darwin'
        ? path.join(this.home, 'Library', 'Caches')
        : null,
    )
    return x
  }
  get requireCachePath() {
    return path.join(this.cacheDir, '/.require-cache.json')
  }
  get requestsCachePath() {
    return path.join(this.cacheDir, '/.requests.json')
  }
  findConfigDir(): null | string {
    const configPath = findUp.sync(
      ['.graphqlconfig', '.graphqlconfig.yml', '.graphqlconfig.yaml'],
      {
        cwd: this.cwd,
      },
    )

    if (configPath) {
      return path.dirname(configPath)
    }

    return null
  }
  private readPackageJson(options?: RunOptions) {
    if (options) {
      this.mock = options.mock
      this.argv = options.argv || this.argv
      if (options.root) {
        this.root = options.root
        const pjsonPath = path.join(options.root, 'package.json')
        const pjson = fs.readJSONSync(pjsonPath)
        if (pjson && pjson['cli-engine']) {
          this.pjson = pjson
          this.version = pjson.version
        }
      }
    } else {
      const root = getRoot()
      this.root = root
      const pjsonPath = path.join(root, 'package.json')
      const pjson = fs.readJSONSync(pjsonPath)
      if (pjson && pjson['cli-engine']) {
        this.pjson = pjson
        this.version = pjson.version
      }
    }
  }
  private setPaths() {
    this.globalPrismaPath = path.join(this.home, '.prisma/')
    this.globalConfigPath = path.join(this.home, '.prisma/config.yml')
    this.globalClusterCachePath = path.join(this.home, '.prisma/cache.yml')
  }
  private warn(msg: string) {
    this.warnings.push(msg)
  }
  private setDefinitionPaths() {
    const definitionPath = path.join(this.cwd, 'prisma.yml')
    const definitionPathWithPrisma = path.join(this.cwd, 'prisma', 'prisma.yml')
    if (fs.pathExistsSync(definitionPath)) {
      this.definitionDir = this.cwd
      this.definitionPath = definitionPath
    } else if (fs.pathExistsSync(definitionPathWithPrisma)) {
      this.definitionDir = path.join(this.cwd, 'prisma')
      this.definitionPath = definitionPathWithPrisma
    } else {
      this.definitionPath = this.getDefinitionPathByGraphQLConfig()
      if (this.definitionPath) {
        this.definitionDir = path.dirname(this.definitionPath)
      } else {
        const found = findUp.sync('prisma.yml', { cwd: this.cwd })
        this.definitionDir = found ? path.dirname(found) : this.cwd
        this.definitionPath = found || null
      }
    }
  }
  private getDefinitionPathByGraphQLConfig(): string | null {
    // try to lookup with graphql config
    let definitionPath
    try {
      const configDir = this.findConfigDir()!
      const config = getGraphQLConfig(configDir).config

      const allExtensions = [
        config.extensions,
        ...values(config.projects).map(p => p.extensions),
      ]

      const prismaExtension = allExtensions.find(e => Boolean(e && e.prisma))
      if (prismaExtension) {
        let { prisma } = prismaExtension
        if (!fs.pathExistsSync(prisma)) {
          prisma = path.join(configDir, prisma)
        }
        definitionPath = path.resolve(prisma)
      }
      this.definitionDir = configDir
    } catch (e) {
      //
    }

    return definitionPath
  }
  private getCwd() {
    // get cwd
    let cwd = process.cwd()
    if (process.env.NODE_ENV === 'test' && process.env.TEST_PRISMA_CLI) {
      cwd = path.join(os.tmpdir(), `${cuid()}/`)
      fs.mkdirpSync(cwd)
    }
    return cwd
  }
  private getHome() {
    // get home
    let home = os.homedir() || os.tmpdir()
    if (process.env.NODE_ENV === 'test') {
      home = path.join(os.tmpdir(), `${cuid()}/`)
      fs.mkdirpSync(home)
    }
    return home
  }
}

function dir(config: Config, category: string, d: string | null): string {
  const cacheKey = `dir:${category}`
  const cache = config.__cache[cacheKey]
  if (cache) {
    return cache
  }
  d =
    d ||
    path.join(
      config.home,
      category === 'data' ? '.local/share' : '.' + category,
    )
  if (config.windows) {
    d = process.env.LOCALAPPDATA || d
  }
  d = process.env.XDG_DATA_HOME || d
  d = path.join(d, config.dirname)
  fs.mkdirpSync(d)
  config.__cache[cacheKey] = d
  return d
}
