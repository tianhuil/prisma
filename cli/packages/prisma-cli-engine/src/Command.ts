import 'source-map-support/register'
import { Arg, Flags } from './Flags/index'
import { Output } from './Output'
import { Config } from './Config'
import { ProjectDefinition, RunOptions } from './types/common'
import { OutputArgs, OutputFlags, Parser } from './Parser'
import Help from './Help'
import { Client } from './Client/Client'
import { Environment, PrismaDefinitionClass } from 'prisma-yml'
import packagejson = require('../package.json')
import * as mock from './mock'
import { RC } from './types/rc'
import { initStatusChecker } from './StatusChecker'
import { filterObject } from './util'
import * as os from 'os'
import * as fs from 'fs-extra'
import * as path from 'path'
import chalk from 'chalk';
const debug = require('debug')('command')

const pjson = packagejson as any

export class Command {
  static topic: string
  static group: string
  static command?: string
  static description?: string
  static usage?: string
  static flags: Flags
  static args: Arg[] = []
  static aliases: string[] = []
  static hidden: boolean = false
  static mockDefinition: ProjectDefinition
  static mockRC: RC
  static allowAnyFlags: boolean = false
  static deprecated?: boolean = false
  static printVersionSyncWarning: boolean = false

  static get id(): string {
    return this.command ? `${this.topic}:${this.command}` : this.topic
  }

  static async mock(...argv: any[]): Promise<Command> {
    let customArgs: any = null
    if (typeof argv[0] === 'object') {
      customArgs = argv.shift()
    }

    argv.unshift('argv0', 'cmd')

    const mockDefinition =
      customArgs && customArgs.mockDefinition
        ? customArgs.mockDefinition
        : mock.mockDefinition
    const mockRC = customArgs && customArgs.mockRC ? customArgs.mockRC : null
    const mockConfig =
      customArgs && customArgs.mockConfig ? customArgs.mockConfig : null
    debug(`Using mockDefinition`, mockDefinition)
    debug(`Using mockRC`, mockRC)

    return this.run({ argv, mock: true, mockDefinition, mockRC, mockConfig })
  }

  printServerVersion(version: string | null) {
    if (version) {
      return `${os.EOL}Prisma server version: ${version}`
    } else {
      return ``
    }
  }

  printCLIVersion() {
    return `Prisma CLI version: ${this.config.userAgent}`
  }

  getVersionTokens(v: string): { minorVersion: string, stage: string } {
    const tokens = v.replace('-', '.').split('.')
    if (tokens.length < 2) {
      throw new Error(`Unable to construct minor version from ${v}`)
    }
    let stage = 'master'
    stage = tokens.some(t => t.includes('beta')) ? 'beta' : 'master'
    if (stage === 'master') {
      stage = tokens.some(t => t.includes('alpha')) ? 'alpha' : 'master'
    }
    return {
      minorVersion: `${tokens[0]}.${tokens[1]}`.replace('-alpha', '').replace('-beta', ''),
      stage: stage
    }
  }

  printVersionSyncWarningMessage() {
    return `${chalk.yellow(chalk.bold(`Warning: `))}Your Prisma server and Prisma CLI are currently out of sync. They should be on the same minor version.`
  }

  compareVersions(v1, v2) {
    const v1Tokens = this.getVersionTokens(v1)
    const v2Tokens = this.getVersionTokens(v2)
    return v1Tokens.minorVersion === v2Tokens.minorVersion && v1Tokens.stage === v2Tokens.stage
  }

  async areServerAndCLIInSync(cmd: Command): Promise<{
    inSync: boolean,
    serverVersion: string | null
  }> {
    try {
      const envFile = cmd.flags['env-file']
      if (envFile && !fs.pathExistsSync(path.join(cmd.config.cwd, envFile))) {
        await cmd.out.error(`--env-file path '${envFile}' does not exist`)
      }
      await cmd.definition.load(cmd.flags, envFile)

      const server = await cmd.definition.getCluster(false)
      if (server) {
        try {
          const serverVersion = await server.getVersion()
          return {
            inSync: this.compareVersions(serverVersion, cmd.config.version),
            serverVersion
          }
        } catch (e) {
          debug(`Failed to fetch server version`)
          debug(e.toString())
          return {
            inSync: true,
            serverVersion: null
          }
        }
      } else {
        debug(`Failed to get the server`)
        return {
          inSync: true,
          serverVersion: null
        }
      }
    } catch(e) {
      debug(`Failed to get the definition file`)
      debug(e.toString())
      return {
        inSync: true,
        serverVersion: null
      }
    }
  }

  static async run(config?: RunOptions): Promise<Command> {
    const cmd = new this({ config })

    try {
      await cmd.init(config)
      
      if (this.printVersionSyncWarning) {
        const { inSync, serverVersion } = await cmd.areServerAndCLIInSync(cmd)
        if (!inSync) {
          cmd.out.log(`${cmd.printVersionSyncWarningMessage()}
  
${cmd.printCLIVersion()}${cmd.printServerVersion(serverVersion)}
  
For further information, please read: http://bit.ly/prisma-cli-server-sync
  `)
        }
      }

      await cmd.run()
      await cmd.out.done()
    } catch (err) {
      cmd.out.error(err)
    }

    return cmd
  }

  static buildHelp(config: Config): string {
    const help = new Help(config)
    return help.command(this)
  }

  static buildHelpLine(config: Config): string[] {
    const help = new Help(config)
    return help.commandLine(this)
  }

  protected static version = pjson.version

  client: Client
  out: Output
  config: Config
  definition: PrismaDefinitionClass
  // auth: Auth
  env: Environment
  flags: OutputFlags
  args?: OutputArgs
  argv: string[]

  constructor(options: { config?: RunOptions } = { config: { mock: false } }) {
    if (options.config && options.config instanceof Config) {
      this.config = options.config
    } else if (options && options.config && options.config.mockConfig) {
      this.config = options.config.mockConfig
    } else {
      this.config = new Config(options.config)
    }
    this.out = new Output(this.config)
    this.config.setOutput(this.out)
    this.argv = options.config && options.config.argv ? options.config.argv : []
    this.env = new Environment(this.config.home, this.out, this.config.version)
    this.definition = new PrismaDefinitionClass(
      this.env,
      this.config.definitionPath,
      process.env,
      this.out,
    )
    this.client = new Client(this.config, this.env, this.out)
    // this.auth = new Auth(this.out, this.config, this.env, this.client)
    // this.client.setAuth(this.auth)
  }

  async run(...rest: void[]): Promise<void> {
    // noop
  }

  async init(options?: RunOptions) {
    const parser = new Parser({
      flags: (this.constructor as any).flags || {},
      args: (this.constructor as any).args || [],
      variableArgs: (this.constructor as any).variableArgs,
      cmd: this,
    })
    const { argv, flags, args } = await parser.parse({
      flags: this.flags,
      argv: this.argv.slice(2),
    })
    this.flags = flags!
    this.argv = argv!
    this.args = args
    await this.env.load()
    initStatusChecker(this.config, this.env)
  }

  get stdout(): string {
    return this.out.stdout.output
  }

  get stderr(): string {
    return this.out.stderr.output
  }

  getSanitizedFlags(): OutputFlags {
    return filterObject(this.flags, (_, value) => {
      if (value === undefined) {
        return false
      }

      return true
    })
  }
}
