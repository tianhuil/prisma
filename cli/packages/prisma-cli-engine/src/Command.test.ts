/* tslint:disable */
import 'source-map-support/register'

import { Command as Base } from './Command'
import { flags as Flags } from './Flags'
import nock from 'nock'
import { Config } from './Config'

class Command extends Base {
  static topic = 'foo'
  static command = 'bar'
  static flags = { myflag: Flags.boolean() }
  static args = [{ name: 'myarg', required: false }]
}

test('gets the version tokens correctly', async () => {
  const cmd = await Command.mock()
  expect(cmd.getVersionTokens('1.30.0')).toEqual({
    minorVersion: '1.30',
    stage: 'master'
  })
  expect(cmd.getVersionTokens('1.31.0-beta')).toEqual({
    minorVersion: '1.31',
    stage: 'beta'
  })
  expect(cmd.getVersionTokens('1.31.0-beta.5')).toEqual({
    minorVersion: '1.31',
    stage: 'beta'
  })
  expect(cmd.getVersionTokens('1.31.2-alpha')).toEqual({
    minorVersion: '1.31',
    stage: 'alpha'
  })
  expect(cmd.getVersionTokens('1.31.0-alpha.5')).toEqual({
    minorVersion: '1.31',
    stage: 'alpha'
  })
  expect(cmd.getVersionTokens('1.31-alpha.5')).toEqual({
    minorVersion: '1.31',
    stage: 'alpha'
  })
  expect(cmd.getVersionTokens('1.31.2-alpha.5')).toEqual({
    minorVersion: '1.31',
    stage: 'alpha'
  })
  expect(cmd.getVersionTokens('1.31-alpha-7')).toEqual({
    minorVersion: '1.31',
    stage: 'alpha'
  })
})

test('compares the versions correctly', async () => {
  const cmd = await Command.mock()
  expect(cmd.compareVersions('1.30.0', '1.29.0')).toBe(false)
  expect(cmd.compareVersions('1.30.2', '1.30.2')).toBe(true)
  expect(cmd.compareVersions('1.30.0', '1.30.2')).toBe(true)
  expect(cmd.compareVersions('1.30.2-alpha', '1.30.0')).toBe(false)
  expect(cmd.compareVersions('1.30.2-alpha', '1.30.0-beta')).toBe(false)
  expect(cmd.compareVersions('1.30.2-alpha', '1.30.5-alpha')).toBe(true)
  expect(cmd.compareVersions('1.30.2-alpha', '1.31.0-alpha')).toBe(false)
  expect(cmd.compareVersions('1.30.2-alpha.5', '1.31.0-alpha.7')).toBe(false)
  expect(cmd.compareVersions('1.31.2-alpha.5', '1.31-alpha.7')).toBe(true)
  expect(cmd.compareVersions('1.31.2-alpha.5', '1.31-alpha-7')).toBe(true)
})

test('shows the ID', () => {
  expect(Command.id).toEqual('foo:bar')
})

test('runs the command', async () => {
  const cmd = await Command.mock()
  expect(cmd.flags).toEqual({})
  expect(cmd.argv).toEqual([])
})

test('has stdout', async () => {
  class Command extends Base {
    static flags = {
      print: Flags.string(),
      bool: Flags.boolean(),
    }
    async run() {
      this.out.stdout.log(this.flags.print)
    }
  }

  const { stdout } = await Command.mock('--print=foo')
  expect(stdout).toEqual('foo\n')
})

test('has stderr', async () => {
  class Command extends Base {
    static flags = { print: Flags.string() }
    async run() {
      this.out.stderr.log(this.flags.print)
    }
  }

  const { stderr } = await Command.mock('--print=foo')
  expect(stderr).toEqual('foo\n')
})

test('parses args', async () => {
  const cmd = await Command.mock('one')
  expect(cmd.flags).toEqual({})
  expect(cmd.argv).toEqual(['one'])
  expect(cmd.args).toEqual({ myarg: 'one' })
})

test('has help', async () => {
  class Command extends Base {
    static topic = 'config'
    static command = 'get'
    static help = `this is

some multiline help
`
  }
  let config = new Config({ mock: true })
  expect(Command.buildHelp(config)).toMatchSnapshot('buildHelp')
  expect(Command.buildHelpLine(config)).toMatchSnapshot('buildHelpLine')
})
