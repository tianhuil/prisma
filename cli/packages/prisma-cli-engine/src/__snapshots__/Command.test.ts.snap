// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`has help: buildHelp 1`] = `
"Usage: prisma config get

this is

some multiline help
"
`;

exports[`has help: buildHelpLine 1`] = `
Array [
  "config get",
  null,
]
`;
